File: android/app/src/test/kotlin/de/godisch/potillus/domain/StatsAggregatorVectorTest.kt
/* vim: set et ts=4:
 * =============================================================================
 * Libellus Potionis - Privacy-Friendly Alcohol Tracker
 * Copyright (c) 2026 Martin A. Godisch <martin@godisch.de>
 * =============================================================================
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 * In addition, as permitted by section 7 of the GNU General Public License,
 * this program may carry additional permissions; any such permissions that
 * apply to it are stated in the accompanying COPYING.md file.
 *
 * =============================================================================
 */
package de.godisch.potillus.domain

// =============================================================================
// StatsAggregatorVectorTest.kt – cross-platform parity suite for the three
// aggregations behind the Statistics screen and the report
// =============================================================================
//
// Asserts `StatsAggregator` against `test-vectors/stats-aggregator.json`, the
// file the iOS suite loads too (`StatsAggregatorVectorTest.swift`). Every
// vector entry carries `utcOffsetSeconds`, so the clock hour is fixed by the
// file and this test is independent of the JVM's default zone.
// =============================================================================

import de.godisch.potillus.domain.model.ConsumptionEntry
import de.godisch.potillus.domain.model.DrinkCategory
import de.godisch.potillus.domain.model.DrinkDefinition
import org.json.JSONArray
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Test

class StatsAggregatorVectorTest {

    private companion object {
        val VECTORS = SharedTestVectors.load("stats-aggregator")
        const val EPS = 1e-9
    }

    @Test
    fun `the three aggregations match the shared vectors`() {
        val cases = VECTORS.getJSONArray("cases")
        (0 until cases.length()).map { cases.getJSONObject(it) }.forEach { case ->
            val description = case.getString("description")
            val drinks = case.getJSONArray("drinks").let { arr ->
                (0 until arr.length()).map { i ->
                    val pair = arr.getJSONArray(i)
                    DrinkDefinition(
                        id = pair.getLong(0),
                        name = "drink ${pair.getLong(0)}",
                        volumeMl = 100,
                        alcoholPercent = 5.0,
                        category = DrinkCategory.valueOf(pair.getString(1)),
                    )
                }
            }
            val entries = case.getJSONArray("entries").let { arr ->
                (0 until arr.length()).map { i -> arr.getJSONObject(i).toEntry() }
            }
            val expected = case.getJSONObject("expected")

            val breakdown = StatsAggregator.categoryBreakdown(entries, drinks)
            val expectedBreakdown = expected.getJSONObject("categoryBreakdown")
            assertEquals("$description: categories", expectedBreakdown.keys().asSequence().toSet(), breakdown.keys.map { it.name }.toSet())
            breakdown.forEach { (category, grams) ->
                assertEquals("$description: ${category.name}", expectedBreakdown.getDouble(category.name), grams, EPS)
            }
            assertDoubles("$description: hourlyGrams", expected.getJSONArray("hourlyGrams"), StatsAggregator.hourlyGrams(entries))
            assertDoubles(
                "$description: hourBucketAverages",
                expected.getJSONArray("hourBucketAverages"),
                StatsAggregator.hourBucketAverages(entries, case.getInt("effectivePeriodDays")),
            )
        }
    }

    private fun assertDoubles(message: String, expected: JSONArray, actual: List<Double>) {
        assertEquals("$message: size", expected.length(), actual.size)
        for (i in 0 until expected.length()) {
            assertEquals("$message[$i]", expected.getDouble(i), actual[i], EPS)
        }
    }

    /** The vector's entry object as a domain entry; the fields the aggregations do not read are filled with placeholders. */
    private fun JSONObject.toEntry() = ConsumptionEntry(
        drinkId = getLong("drinkId"),
        drinkName = "drink ${getLong("drinkId")}",
        volumeMl = 100,
        alcoholPercent = 5.0,
        gramsAlcohol = getDouble("gramsAlcohol"),
        timestampMillis = getLong("timestampMillis"),
        utcOffsetSeconds = getInt("utcOffsetSeconds"),
        logicalDate = "2026-08-15",
    )
}

File: ios/PotillusKit/Tests/PotillusKitTests/StatsAggregatorVectorTest.swift
// vim: set et ts=4:
// =============================================================================
// Libellus Potionis - Privacy-Friendly Alcohol Tracker
// Copyright (c) 2026 Martin A. Godisch <martin@godisch.de>
// =============================================================================
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
// details.
//
// You should have received a copy of the GNU General Public License along with
// this program.  If not, see <https://www.gnu.org/licenses/>.
//
// In addition, as permitted by section 7 of the GNU General Public License,
// this program may carry additional permissions; any such permissions that
// apply to it are stated in the accompanying COPYING.md file.
//
// =============================================================================
import XCTest

@testable import PotillusKit

/// Drives `StatsAggregator.categoryBreakdown`, `hourlyGrams` and
/// `hourBucketAverages` from the shared `stats-aggregator.json` vectors — the
/// file Android's `StatsAggregatorVectorTest.kt` feeds to its new
/// `domain/StatsAggregator.kt`. Every vector entry carries `utcOffsetSeconds`,
/// which is now the only thing that decides an hour: the reader's zone is not a
/// parameter any more.
final class StatsAggregatorVectorTest: XCTestCase {

    func testTheThreeAggregationsMatchTheVectors() throws {
        let vectors = try TestVectors.load("stats-aggregator", as: StatsAggregatorVectors.self)
        for testCase in vectors.cases {
            let drinks = testCase.drinks.map { pair -> DrinkDefinition in
                DrinkDefinition(
                    id: pair.id, name: "drink \(pair.id)", volumeMl: 100, alcoholPercent: 5.0,
                    category: DrinkCategory.from(stored: pair.category)
                )
            }
            let entries = testCase.entries.map { entry -> ConsumptionEntry in
                ConsumptionEntry(
                    drinkId: entry.drinkId, drinkName: "drink \(entry.drinkId)", volumeMl: 100,
                    alcoholPercent: 5.0, gramsAlcohol: entry.gramsAlcohol,
                    timestampMillis: entry.timestampMillis, logicalDate: "2026-08-15",
                    utcOffsetSeconds: entry.utcOffsetSeconds
                )
            }

            let breakdown = StatsAggregator.categoryBreakdown(entries: entries, drinks: drinks)
            XCTAssertEqual(
                Set(breakdown.keys.map(\.rawValue)), Set(testCase.expected.categoryBreakdown.keys),
                "\(testCase.description): categories"
            )
            for (category, grams) in breakdown {
                XCTAssertEqual(
                    grams, testCase.expected.categoryBreakdown[category.rawValue] ?? -1, accuracy: 1e-9,
                    "\(testCase.description): \(category.rawValue)"
                )
            }
            assertDoubles(
                StatsAggregator.hourlyGrams(entries: entries),
                testCase.expected.hourlyGrams, "\(testCase.description): hourlyGrams"
            )
            assertDoubles(
                StatsAggregator.hourBucketAverages(
                    entries: entries, effectivePeriodDays: testCase.effectivePeriodDays
                ),
                testCase.expected.hourBucketAverages, "\(testCase.description): hourBucketAverages"
            )
        }
    }

    private func assertDoubles(_ actual: [Double], _ expected: [Double], _ message: String) {
        XCTAssertEqual(actual.count, expected.count, "\(message): size")
        for (index, value) in expected.enumerated() where index < actual.count {
            XCTAssertEqual(actual[index], value, accuracy: 1e-9, "\(message)[\(index)]")
        }
    }
}

/// `stats-aggregator.json` — the category breakdown, the 24-hour histogram and
/// the eight 3-hour buckets.
struct StatsAggregatorVectors: Decodable {
    let cases: [Case]

    struct Case: Decodable {
        let description: String
        /// Positional `[id, category]` pairs.
        let drinks: [DrinkPair]
        let entries: [Entry]
        let effectivePeriodDays: Int
        let expected: Expected
    }

    struct DrinkPair: Decodable {
        let id: Int64
        let category: String

        init(from decoder: Decoder) throws {
            var container = try decoder.unkeyedContainer()
            id = try container.decode(Int64.self)
            category = try container.decode(String.self)
        }
    }

    struct Entry: Decodable {
        let drinkId: Int64
        let gramsAlcohol: Double
        let timestampMillis: Int64
        let utcOffsetSeconds: Int
    }

    struct Expected: Decodable {
        let categoryBreakdown: [String: Double]
        let hourlyGrams: [Double]
        let hourBucketAverages: [Double]
    }
}
